src/core/drive/page_view.js

```diff
@@ -55,6 +59,10 @@ export class PageView extends View {
     return this.snapshotCache.get(location)
   }
 
+  isPageRefresh(visit) {
+    return visit && this.lastRenderedLocation.href === visit.location.href
```

Comment:
@jorgemanrubia @afcapel does this conditional incorporate re-rendering failed form submissions where the response is `4xx` (like a `400 Bad Request` or `422 Unprocessable entity`)?

When a Form Submission succeeds, [navigator.formSubmissionSucceededWithResponse()](https://github.com/hotwired/turbo/blob/3d48c865938a745b280feb567ff124ebc144dfb1/src/core/drive/navigator.js#L71-L90) gets called, which is ultimately the place where the Visit gets proposed.

Conversely, when the Form Submission fails with a Client Error (status `400..499`), then [navigator.formSubmissionFailedWithResponse](https://github.com/hotwired/turbo/blob/3d48c865938a745b280feb567ff124ebc144dfb1/src/core/drive/navigator.js#L92-L105) gets called. 

`navigator.formSubmissionFailedWithResponse` doesn't have any concept of a `Visit`, and interacts with the View directly.

Reply:
Thanks for the heads up @seanpdoyle. That's a great scenario for morphing indeed. See https://github.com/hotwired/turbo/pull/1061.